apps/etterlatte-beregning/src/main/kotlin/beregning/InntektAvkortingService.kt

```diff
@@ -42,4 +44,63 @@ object InntektAvkortingService {
                 throw RuntimeException("Ugyldig regler for periode: ${resultat.ugyldigeReglerForPeriode}")
         }
     }
+
+    fun beregnAvkortetYtelse(
+        beregnigner: List<Beregningsperiode>,
+        avkortingGrunnlag: List<AvkortingGrunnlag>
+    ): List<AvkortetYtelse> {
+        val avkortingsgrunnlag = periodisertBruttoYtelseOgAvkorting(beregnigner, avkortingGrunnlag)
+        return avkortingsgrunnlag.map { grunnlag ->
+            val resultat = avkorteYtelse.eksekver(grunnlag, RegelPeriode(grunnlag.fom.atDay(1)))
+            when (resultat) {
+                is RegelkjoeringResultat.Suksess -> {
+                    resultat.periodiserteResultater.map { periodisertResultat ->
+                        AvkortetYtelse(
+                            periode = Periode(
+                                fom = YearMonth.from(periodisertResultat.periode.fraDato),
+                                tom = periodisertResultat.periode.tilDato?.let { YearMonth.from(it) }
+                            ),
+                            ytelseEtterAvkorting = periodisertResultat.resultat.verdi,
+                            tidspunkt = Tidspunkt.now(),
+                            regelResultat = periodisertResultat.toJsonNode()
+                        )
+                    }
+                }
+
+                is RegelkjoeringResultat.UgyldigPeriode ->
+                    throw RuntimeException("Ugyldig regler for periode: ${resultat.ugyldigeReglerForPeriode}")
+            }
+        }.flatten()
+    }
+
+    private fun periodisertBruttoYtelseOgAvkorting(
+        beregnigner: List<Beregningsperiode>,
+        avkortingGrunnlag: List<AvkortingGrunnlag>
+    ): List<AvkortetYtelseGrunnlag> {
+        val beregnedeAvkortinger = avkortingGrunnlag.flatMap { it.beregnetAvkorting }
+        val knekkpunkter = beregnigner.map { it.datoFOM }.toMutableSet() + beregnedeAvkortinger.map { it.periode.fom }
+
+        return knekkpunkter.map { knekkpunkt ->
+            val beregning = beregnigner.filter { it.datoFOM <= knekkpunkt }.maxByOrNull { it.datoFOM }
+                ?: throw Exception("Noe gikk galt ved uthenting av grunnlag for avkorting")
+
+            val beregnetAvkorting = beregnedeAvkortinger.filter { it.periode.fom <= knekkpunkt }
+                .maxByOrNull { it.periode.fom }
+                ?: throw Exception("Noe gikk galt ved uthenting av grunnlag for avkorting")
+
+            AvkortetYtelseGrunnlag(
+                fom = knekkpunkt,
+                bruttoYtelse = FaktumNode(
+                    verdi = beregning.utbetaltBeloep,
+                    kilde = "Regel",
```

Comment:
Tror det finnes en slags spesifikk regelkilde - mulig det kan brukes her?

Reply:
Oppretta oppgave for bruk av kilder for regler
https://jira.adeo.no/browse/EY-2123